cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeMonitorClient.java

```diff
@@ -91,6 +104,61 @@ public RuntimeMonitorClient(String hostname, int port, HttpRequestConfig request
     }
   }
 
+  /**
+   * Encode request to avro binary format.
+   * @param topicsToRequest topic requests to be
+   * @param outputStream Outputstream to write to
+   * @throws IOException if there is any exception while reading from output stream
+   */
+  private void encodeRequest(Map<String, MonitorConsumeRequest> topicsToRequest,
+                            OutputStream outputStream) throws IOException {
+    Encoder encoder = EncoderFactory.get().directBinaryEncoder(outputStream, null);
+    encoder.writeMapStart();
+    encoder.setItemCount(topicsToRequest.size());
+
+    DatumWriter<GenericRecord> requestDatumWriter = new GenericDatumWriter<>(
+      MonitorSchemas.V1.MonitorConsumeRequest.SCHEMA.getValueType());
+
+    for (Map.Entry<String, MonitorConsumeRequest> requestEntry : topicsToRequest.entrySet()) {
+      encoder.startItem();
+      encoder.writeString(requestEntry.getKey());
+      requestDatumWriter.write(requestEntry.getValue().toGenericRecord(), encoder);
+    }
+
+    encoder.writeMapEnd();
+  }
+
+  /**
+   * Decodes avro binary response.
+   * @param is input stream to read from
+   * @throws IOException if there is any exception while reading from input stream
+   */
+  private Map<String, Deque<MonitorMessage>> decodeResponse(InputStream is) throws IOException {
+    Decoder decoder = DecoderFactory.get().directBinaryDecoder(is, null);
+    GenericRecord reuse = new GenericData.Record(MonitorSchemas.V1.MonitorResponse.SCHEMA.getValueType()
+                                                   .getElementType());
+
+    Map<String, Deque<MonitorMessage>> decodedMessages = new HashMap<>();
+    long entries = decoder.readMapStart();
+    while (entries > 0) {
```

Comment:
since you are manually decoding map / array in avro, you should keep reading for the next number of entries once the current block is done reading. Please refer to the javadoc of the `Decoder.readMapStart()` and `Decoder.readArrayStart()` for the correct implementation.